### PR TITLE
fix: make TestDB iterator deterministic

### PR DESCRIPTION
**Pull request by blasrodri**
Fixes: #7024

Test plan
---------
I did not find a test that could be updated/enhanced to validate that the iteration would remain deterministic.
Not sure whether this extra step is needed. But if it is, I'd appreciate some guidance.

Edit:

Added two tests: `rocksdb_iter_order` and `testdb_iter_order`. These two tests that there is determinism
when calling the `iter` functions.

### CODE REVIEW

File: core/store/src/db.rs [hunk 1]
```diff
@@ -259,7 +259,7 @@ impl RocksDB {
 }
 
 pub struct TestDB {
-    db: RwLock<enum_map::EnumMap<DBCol, HashMap<Vec<u8>, Vec<u8>>>>,
+    db: RwLock<enum_map::EnumMap<DBCol, BTreeMap<Vec<u8>, Vec<u8>>>>,
```
Comment (mina86) [inline core/store/src/db.rs:262]: Could you add a comment why we’re using BTreeMap i/o a HashMap?

**Review by mina86 — COMMENTED** [review]
And could you also update TestDB::iter_prefix implementation such that it takes advantage of the ordered structure.

**Comment by mina86** [comment]
Also, here’s possible test to add at the end of store/lib.rs:

```rust
    /// Asserts that elements in the vector are sorted.
    fn assert_sorted(want_count: usize, keys: Vec<Box<[u8]>>) {
        assert_eq!(want_count, keys.len());
        for (pos, pair) in keys.windows(2).enumerate() {
            let (fst, snd) = (&pair[0], &pair[1]);
            assert!(fst <= snd, "{fst:?} > {snd:?} at {pos}");
        }
    }

    /// Checks that keys are sorted when iterating.
    fn test_iter_order_impl(store: crate::Store, count: usize) {
        use rand::Rng;

        // An arbitrary non-rc non-insert-only column we can write data into.
        const COLUMN: crate::DBCol = crate::DBCol::Peers;
        assert!(!COLUMN.is_rc());
        assert!(!COLUMN.is_insert_only());

        // Fill column with random keys.  We're inserting three sets of keys.
        // One set prefixed by "foo", second by "bar" and last by "baz".  Each
        // set is `count` keys (for total of `3*count` keys).
        let mut rng: rand::rngs::StdRng = rand::SeedableRng::seed_from_u64(0x3243f6a8885a308d);
        let mut update = store.store_update();
        let mut buf = [0u8; 20];
        for prefix in [b"foo", b"bar", b"baz"] {
            buf[..prefix.len()].clone_from_slice(prefix);
            for _ in 0..count {
                rng.fill(&mut buf[prefix.len()..]);
                update.set(COLUMN, &buf, &buf);
            }
        }
        update.commit().unwrap();

        // Check that full scan produces keys in proper order.
        let keys: Vec<Box<[u8]>> = store.iter(COLUMN).map(|(key, _)| key).collect();
        assert_sorted(3 * count, keys);

        let keys: Vec<Box<[u8]>> = store.iter_raw_bytes(COLUMN).map(|(key, _)| key).collect();
        assert_sorted(3 * count, keys);

        // Check that prefix scan produces keys in proper order.
        let keys: Vec<Box<[u8]>> = store.iter_prefix(COLUMN, b"baz").map(|(key, _)| key).collect();
        for (pos, key) in keys.iter().enumerate() {
            assert_eq!(b"baz", &key[0..3],
                       "Expected ‘baz’ prefix but got {key:?} at {pos}");
        }
        assert_sorted(count, keys);
    }

    #[test]
    fn rocksdb_iter_order() {
        let (_tmp_dir, opener) = crate::Store::tmp_opener();
        test_iter_order_impl(opener.open(), 10_000);
    }

    #[test]
    fn testdb_iter_order() {
        test_iter_order_impl(crate::test_utils::create_test_store(), 10_000);
    }
```

**Comment by RieserStern** [comment]
Hello
I'm interested your project as a Senior blockchain developer, who has enough experience developing

**Comment by RieserStern** [comment]
my discord id: RieserStern#3292

**Comment by blasrodri** [comment]
> Also, here’s possible test to add at the end of store/lib.rs:
> 
> ```rust
>     /// Asserts that elements in the vector are sorted.
>     fn assert_sorted(want_count: usize, keys: Vec<Box<[u8]>>) {
>         assert_eq!(want_count, keys.len());
>         for (pos, pair) in keys.windows(2).enumerate() {
>             let (fst, snd) = (&pair[0], &pair[1]);
>             assert!(fst <= snd, "{fst:?} > {snd:?} at {pos}");
>         }
>     }
> 
>     /// Checks that keys are sorted when iterating.
>     fn test_iter_order_impl(store: crate::Store, count: usize) {
>         use rand::Rng;
> 
>         // An arbitrary non-rc non-insert-only column we can write data into.
>         const COLUMN: crate::DBCol = crate::DBCol::Peers;
>         assert!(!COLUMN.is_rc());
>         assert!(!COLUMN.is_insert_only());
> 
>         // Fill column with random keys.  We're inserting three sets of keys.
>         // One set prefixed by "foo", second by "bar" and last by "baz".  Each
>         // set is `count` keys (for total of `3*count` keys).
>         let mut rng: rand::rngs::StdRng = rand::SeedableRng::seed_from_u64(0x3243f6a8885a308d);
>         let mut update = store.store_update();
>         let mut buf = [0u8; 20];
>         for prefix in [b"foo", b"bar", b"baz"] {
>             buf[..prefix.len()].clone_from_slice(prefix);
>             for _ in 0..count {
>                 rng.fill(&mut buf[prefix.len()..]);
>                 update.set(COLUMN, &buf, &buf);
>             }
>         }
>         update.commit().unwrap();
> 
>         // Check that full scan produces keys in proper order.
>         let keys: Vec<Box<[u8]>> = store.iter(COLUMN).map(|(key, _)| key).collect();
>         assert_sorted(3 * count, keys);
> 
>         let keys: Vec<Box<[u8]>> = store.iter_raw_bytes(COLUMN).map(|(key, _)| key).collect();
>         assert_sorted(3 * count, keys);
> 
>         // Check that prefix scan produces keys in proper order.
>         let keys: Vec<Box<[u8]>> = store.iter_prefix(COLUMN, b"baz").map(|(key, _)| key).collect();
>         for (pos, key) in keys.iter().enumerate() {
>             assert_eq!(b"baz", &key[0..3],
>                        "Expected ‘baz’ prefix but got {key:?} at {pos}");
>         }
>         assert_sorted(count, keys);
>     }
> 
>     #[test]
>     fn rocksdb_iter_order() {
>         let (_tmp_dir, opener) = crate::Store::tmp_opener();
>         test_iter_order_impl(opener.open(), 10_000);
>     }
> 
>     #[test]
>     fn testdb_iter_order() {
>         test_iter_order_impl(crate::test_utils::create_test_store(), 10_000);
>     }
> ```

Thanks! Added!

File: core/store/src/db.rs [hunk 2]
```diff
@@ -259,7 +259,10 @@ impl RocksDB {
 }
 
 pub struct TestDB {
-    db: RwLock<enum_map::EnumMap<DBCol, HashMap<Vec<u8>, Vec<u8>>>>,
+    // In order to ensure determinism when iterating over column's results
+    // a BTreeMap is used since it is an ordered map. A HashMap would
+    // give the aforementioned guarantee, and therefore is discarded.
```
Comment (mina86) [inline core/store/src/db.rs:264]:
```suggestion
    // Use BTreeMap in order to ensure sorted order when iterating over a
    // column.  HashMap would return items in arbitrary order.
```

**Comment by blasrodri** [comment]
> LG, just last thing as I mentioned in previous comment is to make TestDB iter_prefix implementation take advantage of [`BTreeMap::range`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.range) so that it doesn’t scan the entire map.

Oh sorry! I haven't seen it. I'll do that :)

**Comment by blasrodri** [comment]
> 

Can't avoid allocating

```diff
diff --git a/core/store/src/db.rs b/core/store/src/db.rs
index 42597e6c5..fc14d75c7 100644
--- a/core/store/src/db.rs
+++ b/core/store/src/db.rs
@@ -429,10 +429,14 @@ impl Database for TestDB {
         col: DBCol,
         key_prefix: &'a [u8],
     ) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
-        RocksDB::iter_with_rc_logic(
-            col,
-            self.iter(col).filter(move |(key, _value)| key.starts_with(key_prefix)),
-        )
+        let key_prefix = key_prefix.to_vec();
+        let iterator = self.db.read().unwrap()[col]
+            .clone()
+            .range(dbg!(key_prefix.clone())..)
+            .take_while(move |(k, _)| k.clone().starts_with(&key_prefix))
+            .map(|(k, v)| (k.clone().into_boxed_slice(), v.clone().into_boxed_slice()))
+            .collect::<Vec<_>>();
+        RocksDB::iter_with_rc_logic(col, Box::new(iterator.into_iter()))
     }
```

Any suggestions?

File: core/store/src/db.rs [hunk 3]
```diff
@@ -429,10 +429,12 @@ impl Database for TestDB {
...
     ) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
-        RocksDB::iter_with_rc_logic(
-            col,
-            self.iter(col).filter(move |(key, _value)| key.starts_with(key_prefix)),
-        )
+        let iterator = self.db.read().unwrap()[col]
+            .range(key_prefix.to_vec()..)
+            .take_while(move |(k, _)| k.starts_with(&key_prefix))
+            .map(|(k, v)| (k.clone().into_boxed_slice(), v.clone().into_boxed_slice()))
+            .collect::<Vec<_>>();
```
Comment (blasrodri) [inline core/store/src/db.rs:436]: couldn't find out how to do less allocations

Comment (mina86) [reply core/store/src/db.rs:436]: LG, we’re doing even larger allocations in other iter methods.

File: core/store/src/db.rs [hunk 4]
```diff
@@ -426,10 +429,12 @@ impl Database for TestDB {
...
+        RocksDB::iter_with_rc_logic(col, Box::new(iterator.into_iter()))
```
Comment (mina86) [inline core/store/src/db.rs:437]: You don’t need to box this iterator. `iter_with_rc_logic` takes generic iterator.

**Review by mina86 — APPROVED** [review]

File: core/store/src/db.rs (hunk 4 above)
Comment (blasrodri) [reply core/store/src/db.rs]: you're right! Done, thanks!

**Pull request merged**